### PR TITLE
gui: Defer removeAndDeleteWallet when no modal widget is active

### LINKED ISSUE #15310: gui: crash if encrypt / change passphrase window is open and wallet is unloaded
Using master @ cb35f1d305d88934df64c2e7fb80700b673360e6

Run `bitcoin-qt`, and open either the `Encrypt Wallet` or `Change Passphrase` window. 
Unload the wallet. ie `src/bitcoin-cli unloadwallet`

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff5ba9523e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff5bb4bc1c libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff5b9fe1c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff5bb0d6e2 libsystem_malloc.dylib`malloc_vreport + 545
    frame #4: 0x00007fff5bb0d4a3 libsystem_malloc.dylib`malloc_report + 152
    frame #5: 0x0000000101ef9e90 QtCore`QObjectPrivate::deleteChildren() + 224
    frame #6: 0x00000001012439e2 QtWidgets`QWidget::~QWidget() + 1058
    frame #7: 0x00000001000ddc8c bitcoin-qt`WalletView::~WalletView() [inlined] WalletView::~WalletView(this=0x0000000159d976e0) at walletview.cpp:87 [opt]
    frame #8: 0x00000001000ddc87 bitcoin-qt`WalletView::~WalletView() [inlined] WalletView::~WalletView(this=0x0000000159d976e0) at walletview.cpp:86 [opt]
    frame #9: 0x00000001000ddc87 bitcoin-qt`WalletView::~WalletView(this=0x0000000159d976e0) at walletview.cpp:86 [opt]
    frame #10: 0x00000001000d42f8 bitcoin-qt`WalletFrame::removeWallet(this=0x000000010eaa4130, wallet_model=<unavailable>) at walletframe.cpp:98 [opt]
    frame #11: 0x000000010001f5b6 bitcoin-qt`BitcoinGUI::removeWallet(this=0x0000000108dc8bb0, walletModel=<unavailable>) at bitcoingui.cpp:611 [opt]
    frame #12: 0x0000000101f01d7b QtCore`QMetaObject::activate(QObject*, int, int, void**) + 2219
    frame #13: 0x00000001000ef3ff bitcoin-qt`WalletController::walletRemoved(this=<unavailable>, _t1=<unavailable>) at moc_walletcontroller.cpp:206 [opt]
    frame #14: 0x00000001000d2f46 bitcoin-qt`WalletController::removeAndDeleteWallet(this=0x0000000159d31a00, wallet_model=0x0000000122f44f80) at walletcontroller.cpp:91 [opt]
    frame #15: 0x0000000101f01d7b QtCore`QMetaObject::activate(QObject*, int, int, void**) + 2219
    frame #16: 0x0000000101efa801 QtCore`QObject::event(QEvent*) + 753
    frame #17: 0x0000000101218f8d QtWidgets`QApplicationPrivate::notify_helper(QObject*, QEvent*) + 269
    frame #18: 0x000000010121a392 QtWidgets`QApplication::notify(QObject*, QEvent*) + 594
```

### PR DESCRIPTION
**Pull request by promag**
Fixes #15310.

### CODE REVIEW

**Comment by promag** [comment]
Alternative to #15313 in 0.18, more details in http://www.erisian.com.au/bitcoin-core-dev/log-2019-03-16.html (line 279) cc @gmaxwell @jonasschnelli

**Comment by jonasschnelli** [comment]
Concept ACK for 0.18.

Looks like a hack (but a good one)... I think we should merge this into 0.18 and properly remove synchronous dialogs in 0.19.

**Comment by jonasschnelli** [comment]
Hmm... second thought: I'm unsure about this. This waits until the users closes the dialogs (rather then closing them explicit). This may lead to an RPC timeout in `unload wallet`...

Maybe we combine this with #15313?

**Comment by promag** [comment]
I don't think that's a problem. The RPC client can timeout (can also happen for other reasons, rescanning for instance?) but the RPC handler still waits. This fixes a unlikely crash (I mean, is there a good reason to do this?) without major changes and can be reverted easily if the future master refactor is backport.

**Comment by promag** [comment]
I'll verify if `QWindow::activeChanged` can be used instead of `QApplication::focusChanged`.

**Comment by MarcoFalke** [comment]
Changed base, needs commit cherry-picked on some other base as well, obviously.

**Comment by promag** [comment]
Rebased.

**Comment by MarcoFalke** [comment]
Should update OP with something like "Fixes #15310"?

**Comment by gmaxwell** [comment]
This seems like a better workaround to the issue for the short term.

**Comment by promag** [comment]
@MarcoFalke yes, after rebased with master.

**Comment by MarcoFalke** [comment]
tested ACK f33e9bcda8a34385872555c1857bbabdf5c68134 (No longer crashes and having the modal gui dialog open will block the unloadwallet RPC, didn't review code too closely)

File: src/qt/walletcontroller.cpp [hunk 1]
```diff
@@ -97,7 +98,16 @@ WalletModel* WalletController::getOrCreateWallet(std::unique_ptr<interfaces::Wal
     m_wallets.push_back(wallet_model);
 
     connect(wallet_model, &WalletModel::unload, [this, wallet_model] {
-        removeAndDeleteWallet(wallet_model);
+        // Defer removeAndDeleteWallet when no modal widget is active.
+        if (QApplication::activeModalWidget()) {
+            connect(qApp, &QApplication::focusChanged, wallet_model, [this, wallet_model]() {
```
Comment (promag) [inline src/qt/walletcontroller.cpp]: Just noting that the 3rd parameter `wallet_model` is used as the destination context, which means:
 - the connection is destroyed when the `wallet_model` is destroyed
 - the connection runs on the wallet model thread (GUI thread) due to `Qt::QueuedConnection`.

**Review by luke-jr — APPROVED** [review]
I suspect there's some race conditions here, but strict improvement over the current condition, so utACK

**Comment by promag** [comment]
@luke-jr care to elaborate?

**Comment by jonasschnelli** [comment]
Tested ACK a10972bc03e1ebbe79c3edc929c91ed53b391c9a

**Comment by luke-jr** [comment]
@promag 

line 104 evaluates to false, then a problematic modal window is opened before line 111

or

line 104 evaluates to true, then the problematic modal window is closed before focusWindowChanged is connected; wallet closing then waits for some arbitrarily future time

**Comment by promag** [comment]
@luke-jr it's impossible, both callbacks runs in the GUI thread and so:
 - first case, no modal can be opened;
 - second case, the modal can't be closed before connecting the signal.

**Comment by luke-jr** [comment]
I thought the whole point of the Qt::QueuedConnection was to move the callback to the GUI thread later?

**Comment by luke-jr** [comment]
Ah, nevermind, I get it - the focusWindowChanged might be in another thread.

**Pull request merged**